create org.osbuild.bootupd stage

dustymabe opened the pull request:
Add the bootupd stage to install GRUB on both BIOS and UEFI systems,
ensuring that your bootloader stays up-to-date.

Signed-off-by: Renata Ravanelli <rravanel@redhat.com>

dustymabe commented:
The 2nd commit here will be dropped. The 3rd and 4th commits are optional and stem from a [previous conversation](https://github.com/osbuild/osbuild/pull/1475#discussion_r1429789726) between @achilleas-k @cgwalters and I. Over there I don't think we fully fleshed out if the optimization was completely out of the question or not. Maybe I should just open a separate PR with that optimization so we could focus discussion on it, but at the same time I don't want to introduce options to the `bootupd` stage here if we'd remove them later.

achilleas-k commented:
> Maybe I should just open a separate PR with that optimization so we could focus discussion on it, but at the same time I don't want to introduce options to the `bootupd` stage here if we'd remove them later.

That would be preferable, yeah.  Can you open it separately so we can talk about that there and then keep this specific to bootupd?  Thanks!

dustymabe commented:
I broke out the controversial commits into https://github.com/osbuild/osbuild/pull/1527

You can now review just the first commit in this PR. The 2nd commit is just there to make tests pass until we can update the registry.gitlab.com/redhat/services/products/image-builder/ci/images/fedora-coreos next week when a new build of FCOS is out that we can use.

ondrejbudai commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+  "type": "array"
+},
+"options": {
+"additionalProperties": true,
```
I suggest the `options` being indented.

dustymabe replied:
I will make that change on the next upload! Thanks for catching that!

dustymabe commented:
@achilleas-k can you pull `quay.io/fedora/fedora-coreos:testing` and push it to `registry.gitlab.com/redhat/services/products/image-builder/ci/images/fedora-coreos:testing` and I'll update this so tests pass and we can hopefully merge.

You can delete `registry.gitlab.com/redhat/services/products/image-builder/ci/images/fedora-coreos:stable` if you want as it should be unused now.

achilleas-k commented:
> @achilleas-k can you pull `quay.io/fedora/fedora-coreos:testing` and push it to `registry.gitlab.com/redhat/services/products/image-builder/ci/images/fedora-coreos:testing` and I'll update this so tests pass and we can hopefully merge.


Done: https://gitlab.com/redhat/services/products/image-builder/ci/images/container_registry/5710257

dustymabe commented:
> > @achilleas-k can you pull `quay.io/fedora/fedora-coreos:testing` and push it to `registry.gitlab.com/redhat/services/products/image-builder/ci/images/fedora-coreos:testing` and I'll update this so tests pass and we can hopefully merge.
> 
> Done: https://gitlab.com/redhat/services/products/image-builder/ci/images/container_registry/5710257

Thanks! Pushed a new update and rebased on top of latest `main`. Hopefully the tests pass!

mvo5 commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+    # risk any inconsistencies with the build root). Let's set up
+    # and chroot to run the bootupctl command from the target.
+    submounts = ['dev', 'proc', 'sys', 'run', 'var', 'tmp']
+    for mnt in submounts:
```
I wonder if we could use a context manager here

dustymabe replied:
SGTM

mvo5 reviewed (commented):
Thank you for this PR - this looks fine but I would really like some tests. When looking at that I got a bit carried away and opened https://github.com/dustymabe/osbuild/pull/14 for your consideration. It addresses the nitpicks/ideas I have in here. I would also love to have a stage test and will look at this next. 

Fwiw, we are super happy that you work on this because we will most likely use it for the bootc container builds :)

dustymabe commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+  "type": "array"
+},
+"options": {
+  "additionalProperties": true,
```
honestly probably because I was copying stuff from another stage? I guess `additionalProperties` just allows for unknown to the schema properties to be specified?

mvo5 replied:
Thanks, yeah, it would unknown properties (which is sometimes what we want and sometimes not :)

dustymabe commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+  "additionalProperties": true,
+  "properties": {
+    "deployment": {
+      "additionalProperties": false,
```
👍

dustymabe commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+      "type": "boolean"
+    },
+    "bios": {
+      "additionalProperties": false,
```
👍

dustymabe commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+        cmd.append(mounts)
+        subprocess.run(cmd, check=True)
+    finally:
+        for mnt in submounts:
```
👍

mvo5 reviewed (dismissed):
This looks good to me now, given that I changed this a bit it might be best if e.g. @achilleas-k could have a look too. 

I would love to have a stage test as well to run this for real and I started to look into it but I think it's a bit involved (because of the required mount setup). I think it's okay to not block on this because there is no regression risk.

mvo5 reviewed (dismissed):
Still looks good!

achilleas-k commented on stages/org.osbuild.bootupd:
```diff
@@ -0,0 +1,125 @@
+      "additionalProperties": false,
+      "required": ["disk"],
+      "properties": {
+        "disk": {
```
NITPICK: Since this option maps to a "device" attached to the stage, perhaps we should call it `device` instead of `disk`?  The description could also be updated to mention that it needs to match the name of a device:
```
Name of stage device to install GRUB for BIOS-based systems.
```
Alternatively, we can force the device name in the schema for the `devices` and omit this (like we do in the `mkfs` stages).  Unless we expect that we might ever have multiple devices.

dustymabe replied:
new upload

achilleas-k commented on stages/org.osbuild.bootupd:
```diff
@@ -25,9 +25,10 @@ SCHEMA_2 = r"""
 },
 "options": {
-  "additionalProperties": true,
+  "additionalProperties": false,
```
Just a git history nitpick:
Should we squash this with the commit that introduced the stage?
Same for L31 below (`"type": "object"`).

dustymabe replied:
new upload

achilleas-k reviewed (changes requested):
1. The last two commits have subjects that are too long.  Can we keep them under 72 chars?
2. Can we also clean up the history a bit? (see comment below).
3. The `disk` option under `bios`: Making enforcing the name through the schema for the `devices` object would make it consistent with how we do it in other stages, but if we prefer to keep it explicit, I'd change the name of the option to make it clear that it maps to a "device".

dustymabe commented:
> 1. The last two commits have subjects that are too long.  Can we keep them under 72 chars?

Fixed.

> 
> 2. Can we also clean up the history a bit? (see comment below).

Tried to fix.

> 
> 3. The `disk` option under `bios`: Making enforcing the name through the schema for the `devices` object would make it consistent with how we do it in other stages, but if we prefer to keep it explicit, I'd change the name of the option to make it clear that it maps to a "device".

I was learning your way on this, but then Renata and I found a gap in
our coverage here (see the newest commit) where on ppc64le we need to
be able to install into a partition of a device so I added the
`partition` key too. Otherwise we could have made `bios: true/false` a
boolean.

achilleas-k commented:
> I was learning your way on this, but then Renata and I found a gap in
> our coverage here (see the newest commit) where on ppc64le we need to
> be able to install into a partition of a device so I added the
> `partition` key too. Otherwise we could have made `bios: true/false` a
> boolean.

You could specify the `start` and `size` options of the device to create a loopback that points to the specific partition.  In the test case for example it look like this:
```json
        {
          "type": "org.osbuild.bootupd",
          "options": {
            "bios": true,
            "static-configs": true,
            "deployment": {
              "osname": "fedora-coreos",
              "ref": "ostree/1/1/0"
            }
          },
          "devices": {
            "disk": {
              "type": "org.osbuild.loopback",
              "options": {
                "filename": "disk.img",
                "start": 4096,
                "size": 260096
              }
            }
          }
        }
```

Would that work?

dustymabe commented:
> Would that work?

unfortunately not. the `grub2-install` stuff really doesn't like it when you map in a slice of a device and not the whole device itself :(

achilleas-k commented:
> > Would that work?
> 
> unfortunately not. the `grub2-install` stuff really doesn't like it when you map in a slice of a device and not the whole device itself :(

Right, of course -_-

achilleas-k reviewed (approved):
LGTM.  Good to go on my end!

dustymabe commented:
Thank you @mvo5 and @achilleas-k for the help and the reviews here.
